network/topology/cache_test.go

```diff
@@ -52,45 +47,89 @@ func TestCache_GenerateFanout_HappyPath(t *testing.T) {
 func TestCache_GenerateFanout_Error(t *testing.T) {
 	// mocks underlying topology returning a fanout
 	top := &mocknetwork.Topology{}
-	log := zerolog.New(os.Stderr).Level(zerolog.DebugLevel)
 	ids := unittest.IdentityListFixture(100)
-	top.On("GenerateFanout", ids).Return(nil, fmt.Errorf("fanout-generation-error")).Once()
+	top.On("GenerateFanout", mock.Anything, mock.Anything).Return(nil, fmt.Errorf("fanout-generation-error")).Once()
 
 	// assumes cache holding some fanout
-	cache := NewCache(log, top)
+	cache := NewCache(zerolog.Nop(), top)
 	cache.cachedFanout = ids.Sample(10)
-	cache.fingerprint = unittest.IdentifierFixture()
+	cache.idsFP = unittest.IdentifierFixture()
+	cache.chansFP = unittest.IdentifierFixture()
 
 	// returning error on fanout generation should invalidate cache
 	// same error should be returned.
-	fanout, err := cache.GenerateFanout(ids)
+	fanout, err := cache.GenerateFanout(ids, network.ChannelList{})
 	require.Error(t, err)
 	require.Nil(t, fanout)
-	require.Equal(t, cache.fingerprint, flow.Identifier{})
+	require.Equal(t, cache.idsFP, flow.Identifier{})
+	require.Equal(t, cache.chansFP, flow.Identifier{})
 	require.Empty(t, cache.cachedFanout)
 
 	// underlying topology should be called only once.
 	mock.AssertExpectationsForObjects(t, top)
 }
 
-// TestCache_Update evaluates that if input to GenerateFanout changes, the cache is invalidated and updated.
-func TestCache_Update(t *testing.T) {
+// TestCache_InputChange_IDs evaluates that if identity list input to GenerateFanout changes,
+// the cache is invalidated and updated.
+func TestCache_InputChange_IDs(t *testing.T) {
 	// mocks underlying topology returning a fanout
 	top := &mocknetwork.Topology{}
-	log := zerolog.New(os.Stderr).Level(zerolog.DebugLevel)
 	ids := unittest.IdentityListFixture(100)
 	fanout := ids.Sample(10)
-	top.On("GenerateFanout", ids).Return(fanout, nil).Once()
+	channels := network.ChannelList{"channel1", "channel2"}
+	top.On("GenerateFanout", mock.Anything, channels).Return(fanout, nil).Once()
 
-	// assumes cache holding some fanout
-	cache := NewCache(log, top)
+	// assumes cache holding some fanout for the same
+	// ids and channels defined above.
+	cache := NewCache(zerolog.Nop(), top)
 	cache.cachedFanout = ids.Sample(10)
-	cache.fingerprint = unittest.IdentifierFixture()
+	cache.idsFP = ids.Fingerprint()
```

Comment:
instead of setting the private member variables of the `cache`, I would rather call `GenerateFanout` once here to more closely resemble to actual usage of the cache.

Reply:
Right, but then the `Cache` test would be all dependent on the underlying topology implementation. I would keep the current test at the level of a general `mock` topology interface, and set the stage. Since what we are testing is _regardless _ of the underlying topology behavior.

Reply:
actually, it wouldn't be dependent on the underlying topology implementation and you have already (and rightly so) mocked it out on line 76. On line 102, you even check the the underlying topology was only called once.
i am nitpicking here and this is I guess just a personal preference, but here you when you set `cache.cachedFanout` you in a way assume the underlying implementation.

Reply:
i was thinking cache.GenerateFanout resulting in a miss, then cache.GenerateFanout with a different list would be a good way to test a cache miss.
But what you have done here also looks good 👍